File: spatial_filtering.m
% Image Smoothing using Mean and Gaussian Filters
function spatial_filtering()
    % Read the image
    % Replace 'peppers.png' with your own image if needed
    % 'peppers.png' is a built-in image in MATLAB
    try
        img = imread('peppers.png');
    catch
        disp('Error: Could not load image. Make sure the image path is correct.');
        return;
    end
    
    % 1. Mean Filter (Averaging Filter)
    % Create a 5x5 average filter (convolution mask)
    mean_filter = fspecial('average', [5 5]);
    
    % Apply the filter using convolution (imfilter)
    % 'replicate' handles border pixels by replicating the nearest border pixel
    mean_filtered = imfilter(img, mean_filter, 'replicate');
    
    % 2. Gaussian Filter
    % Create a 5x5 Gaussian filter with standard deviation (sigma) of 2
    gaussian_filter = fspecial('gaussian', [5 5], 2);
    
    % Apply the filter using convolution
    gaussian_filtered = imfilter(img, gaussian_filter, 'replicate');
    
    % Display the original and filtered images
    figure('Name', 'Spatial Filtering: Noise Reduction', 'Position', [100, 100, 1200, 400]);
    
    subplot(1, 3, 1);
    imshow(img);
    title('Original Image');
    
    subplot(1, 3, 2);
    imshow(mean_filtered);
    title('Mean Filter (5x5)');
    
    subplot(1, 3, 3);
    imshow(gaussian_filtered);
    title('Gaussian Filter (5x5, \sigma=2)');
    
    % Optional: Save the smoothed images to disk
    % imwrite(mean_filtered, 'mean_filtered_matlab.jpg');
    % imwrite(gaussian_filtered, 'gaussian_filtered_matlab.jpg');
    
    disp('Filtering complete. Check the generated figure window.');
end

File: spatial_filtering.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

def apply_smoothing_filters(image_path):
    """
    Applies Mean and Gaussian spatial filters to an image to reduce noise.
    """
    # Load the image
    img = cv2.imread(image_path)
    if img is None:
        print(f"Error: Could not load image at '{image_path}'. Please check the path.")
        return
    
    # Convert from BGR (OpenCV default) to RGB for correct displaying with matplotlib
    img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    
    # 1. Apply Mean Filter (Averaging)
    # Using a 5x5 convolution mask/kernel
    kernel_size = (5, 5)
    mean_filtered = cv2.blur(img_rgb, kernel_size)
    
    # 2. Apply Gaussian Filter
    # Using a 5x5 kernel, sigmaX=0 (calculated automatically from kernel size)
    gaussian_filtered = cv2.GaussianBlur(img_rgb, kernel_size, 0)
    
    # Plotting the results
    plt.figure(figsize=(15, 5))
    
    plt.subplot(1, 3, 1)
    plt.imshow(img_rgb)
    plt.title('Original Image')
    plt.axis('off')
    
    plt.subplot(1, 3, 2)
    plt.imshow(mean_filtered)
    plt.title('Mean Filter (5x5 Kernel)')
    plt.axis('off')
    
    plt.subplot(1, 3, 3)
    plt.imshow(gaussian_filtered)
    plt.title('Gaussian Filter (5x5 Kernel)')
    plt.axis('off')
    
    plt.tight_layout()
    plt.show()

if __name__ == "__main__":
    # Create a dummy image for demonstration if you don't have one
    # Alternatively, replace 'sample.jpg' with a real image path
    try:
        apply_smoothing_filters("sample.jpg")
    except Exception as e:
        print("Make sure 'sample.jpg' exists in the current directory.")
